topdown/eval.go

```diff
@@ -2892,6 +2910,172 @@ func (e evalTerm) save(iter unifyIterator) error {
 
 }
 
+type evalEvery struct {
+	e         *eval
+	expr      *ast.Expr
+	generator ast.Body
+	body      ast.Body
+}
+
+func (e evalEvery) eval(iter unifyIterator) error {
+	// unknowns in domain or body: save the expression, PE its body
+	if e.e.unknown(e.generator, e.e.bindings) || e.e.unknown(e.body, e.e.bindings) {
+		return e.save(iter)
+	}
+
+	child := e.e.closure(e.generator)
+	all := true // all generator evaluations yield one successful body evaluation
+	err := child.Run(func(child *eval) error {
+		if !all {
+			// NOTE(sr): Is this good enough? We don't have a "fail EE".
+			// This would do extra work, like iterating needlessly if domain was a large array.
+			return nil
+		}
+		body := child.closure(e.body)
+		body.findOne = true
+		done := false
+		err := body.eval(func(*eval) error {
+			done = true
+			return nil
+		})
+		if !done {
+			all = false
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	if all {
+		return iter()
+	}
+	return nil
+}
+
+func (e *evalEvery) save(iter unifyIterator) error {
+	// TODO(sr): check traces
+	kvVars := e.expr.Terms.(*ast.Every).KeyValueVars()
+	vars := make([]*ast.Term, 0, len(kvVars))
+	for _, v := range kvVars.Sorted() {
+		vars = append(vars, ast.NewTerm(v))
+	}
+
+	if !e.e.inliningControl.shallow && !e.e.unknown(e.generator, e.e.bindings) {
+		return e.saveUnrolled(vars, kvVars, iter)
+	}
+
+	return e.saveEveryExpr(vars, kvVars, iter)
+}
+
+func (e *evalEvery) saveUnrolled(vars []*ast.Term, kvVars ast.VarSet, iter unifyIterator) error {
+
+	child := e.e.closure(e.generator)
+	e.e.saveSet.Push(vars, child.bindings)
```

Comment:
I'm surprised that we're putting the k/v variables into the save-set because that means they're treated as unknowns. If we're unrolling `every`, I would have expected them to be known.